### ISSUE TITLE
ref struct specification needs lots of work

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#37841]
The specification for ref struct types at
https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/span-safety.md
does not say that they cannot be used in async or iterator methods.  But our translation strategy for such types means they cannot be hoisted to a state machine because they cannot be declared as a field. Similarly, they cannot be pending on the evaluation stack when an await occurs because they cannot be spilled.  These language restrictions need to be documented.

While we're at it it would be nice to have a grammar rule for declaring a ref struct and actual language rules that ensure that they never appear on the heap (e.g. no using such a type for a field except in a ref struct, no using it as a type argument, etc).
